commands.go

```diff
@@ -230,22 +242,33 @@ func testProgram(c *cli.Context) {
 				fmt.Print(string(out))
 			}
 		}
+	} else {
+		cprintf(red, bold, "Compilation Error\n\n")
+		os.Exit(1)
 	}
 
 	run := renderCmd(config.Test[ext].Run, file)
 	var answer string
-	if inputFile := c.String("i"); inputFile == "" {
-		// get test case from udebug.com
-		var input string
-		input, answer = getTestData(pid)
-		run.Stdin = strings.NewReader(input)
-	} else {
-		f, err := os.Open(inputFile)
-		if err != nil {
-			panic(err)
+
+	if run != nil {
+
+		if inputFile := c.String("i"); inputFile == "" {
+			// get test case from udebug.com
+			var input string
+			input, answer = getTestData(pid)
+			run.Stdin = strings.NewReader(input)
```

Comment:
The "Invalid memory address or nil pointer dereference" error is thrown here when we try to access `Stdin` from `nil`, which is returned by `renderCmd`. I suggest let `renderCmd` function panic if it can't construct the command, instead of returning `nil`.

Also `config.Test` is of type map and can directly be indexed by `ext`, no "loop and if` check needed.